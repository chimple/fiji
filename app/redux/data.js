const FETCH_GAME_DATA_REQUEST = 'Fiji/game/FETCH_GAME_DATA_REQUEST'
const FETCH_GAME_DATA_SUCCESS = 'Fiji/game/FETCH_GAME_DATA_SUCCESS'
const FETCH_GAME_DATA_FAILURE = 'Fiji/game/FETCH_GAME_DATA_FAILURE'

export const initialState = {
  isFetching: false,
  gameData: []
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case FETCH_GAME_DATA_REQUEST:
      return {
        ...state,
        gameData: [],
        isFetching: true
      }
    case FETCH_GAME_DATA_SUCCESS:
      return {
        ...state,
        isFetching: false,
        gameData: action.gameData
      }
    case FETCH_GAME_DATA_FAILURE:
      return {
        ...state,
        isFetching: false,
        gameData: []
      }
    default:
      return state
  }
}

export const fetchGameDataRequest = () => ({
  type: FETCH_GAME_DATA_REQUEST
})

export const fetchGameDataSuccess = (gameData) => ({
  type: FETCH_GAME_DATA_SUCCESS,
  gameData
})

export const fetchGameDataFailure = () => ({
  type: FETCH_GAME_DATA_FAILURE
})

export const fetchMultipleChoiceData = (setId, numChoices, numRows) => (dispatch, getState) => {
  let choices = []
  for (let index = 0; index < 26; index++) {
    choices.push(String.fromCharCode(65 + index))
  }
  dispatch(fetchGameDataRequest())
  let data = []
  for (let i = 0; i < numRows; i++) {
    let entries = []
    while (entries.length < numChoices) {
      let rIndex = getRandomInt(0, choices.length)
      if (!entries.includes(choices[rIndex])) {
        entries.push(choices[rIndex])
      }
    }
    let questionIndex = getRandomInt(0, numChoices)
    data.push({
      question: entries[questionIndex],
      answerIndex: questionIndex,
      choices: entries
    })
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchMatchData = (setId, numSets, numRows) => (dispatch, getState) => {
  let choices = []
  for (let index = 0; index < 26; index++) {
    choices.push(String.fromCharCode(65 + index))
    choices.push(String.fromCharCode(97 + index))
  }
  dispatch(fetchGameDataRequest())
  let data = []
  for (let i = 0; i < numRows; i++) {
    let entries = []
    while (entries.length < numSets) {
      let rIndex = getRandomInt(0, choices.length)
      if (!entries.includes(choices[rIndex])) {
        entries.push(choices[rIndex])
      }
    }
    data.push(entries.map((val) => {
      return [val, val]
    }))
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchSerialData = (setId, numRows) => (dispatch, getState) => {
  let data = []
  dispatch(fetchGameDataRequest())
  for (let index = 0; index < numRows; index++) {
    let answer = getRandomInt(0, 999)
    let start = getRandomInt(2, 10)
    let end = getRandomInt(2, 10)
    let serial = []
    for (let seq = answer - start; seq < answer + end; seq++) {
      serial.push(seq.toString())
    }
    data.push({
      answer,
      serial
    })
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchConsecutiveData = (setId, dataLength, otherLength, numRows) => (dispatch, getState) => {
  let data = []
  const MAX_INT = 999
  dispatch(fetchGameDataRequest())
  for (let index = 0; index < numRows; index++) {
    let start = getRandomInt(1, MAX_INT)
    let serial = []
    for (let seq = start; seq < start + dataLength; seq++) {
      serial.push(seq.toString())
    }
    let others = []
    for (let other = 0; other < otherLength; other++) {
      others.push(((start + dataLength - 1 + getRandomInt(1, MAX_INT - dataLength)) % MAX_INT).toString())
    }
    data.push({
      serial,
      others
    })
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchWordData = (setId, dataLength, otherLength, numRows) => (dispatch, getState) => {
  const choices = [
    ['A'],
    ['I', 'N'],
    ['D', 'O', 'G'],
    ['G', 'I', 'R', 'L'],
    ['A', 'P', 'P', 'L', 'E'],
    ['B', 'A', 'T', 'M', 'A', 'N'],
    ['G', 'I', 'R', 'A', 'F', 'F', 'E'],
    ['B', 'L', 'I', 'Z', 'Z', 'A', 'R', 'D']
  ]
  let letters = []
  for (let index = 0; index < 26; index++) {
    letters.push(String.fromCharCode(65 + index))
  }
  dispatch(fetchGameDataRequest())
  let data = []
  if (dataLength >= 1 && dataLength <= 8) {
    for (let index = 0; index < numRows; index++) {
      let others = []
      while (others.length < otherLength) {
        let rIndex = getRandomInt(0, letters.length)
        if (!choices[dataLength - 1].includes(letters[rIndex])) {
          others.push(letters[rIndex])
        }
      }
      data.push({
        word: choices[dataLength - 1],
        others
      })
    }
    dispatch(fetchGameDataSuccess(data))
  }
}

export const fetchEquationData = (setId, maxDigits, numRows) => (dispatch, getState) => {
  let data = []
  const MAX_INT = maxDigits * 10 - 1
  dispatch(fetchGameDataRequest())
  for (let index = 0; index < numRows; index++) {
    const answer = getRandomInt(1, MAX_INT)
    const firstNum = getRandomInt(1, answer)
    const secondNum = answer - firstNum
    const op = '+'
    data.push({
      firstNum,
      op,
      secondNum,
      answer
    })
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchCrosswordData = (setId, matrixSize, numRows) => (dispatch, getState) => {
  let data = []
  dispatch(fetchGameDataRequest())
  for (let index = 0; index < numRows; index++) {
    data.push({
      crossword: [
        ['E',null,null,null,null],
        ['A',null,null,null,null],
        ['T','I','G','E','R'],
        [null,null,null,null,'A'],
        [null,null,null,null,'T']
      ],
      start: [
        ['image',[0,0],'down'],
        ['image',[2,0],'lToR'],
        ['image',[2,4],'down']        
      ]
    })
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchTrueOrFalseData = (setId, numRows) => (dispatch, getState) => {
  let data = []
  dispatch(fetchGameDataRequest())
  for (let index = 0; index < numRows; index++) {
    data.push({
      question: 'Apple is a:',
      answer: 'Fruit',
      wrongChoice: 'Animal'
    })
  }
  dispatch(fetchGameDataSuccess(data))
}

export const fetchRollingData = (setId, maxWordSize, numChoices, numRows) => (dispatch, getState) => {
  let data = []
  dispatch(fetchGameDataRequest())
  for (let index = 0; index < numRows; index++) {
    data.push({
      question: 'Fruit',
      answer: ['A','P','P','L','E'],
      choices: [
        ['X','Y','Z','A','B','C','D'],
        ['M','N','O','P','Q','R','S'],
        ['M','N','O','P','Q','R','S'],
        ['I','J','K','L','M','N','O'],
        ['B','C','D','E','F','G','H']
      ]
    })
  }
  dispatch(fetchGameDataSuccess(data))
}


getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; //The maximum is exclusive and the minimum is inclusive
}

export default reducer