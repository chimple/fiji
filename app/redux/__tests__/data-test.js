import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'

import data, {
  initialState,
  fetchGameDataRequest,
  fetchGameDataSuccess,
  fetchGameDataFailure,
  fetchMultipleChoiceData,
  fetchMatchData,
  fetchSerialData,
  fetchConsecutiveData,
  fetchWordData,
  fetchEquationData,
  fetchCrosswordData,
  fetchTrueOrFalseData,
  fetchRollingData
} from '../data'

// test reducers

it('returns the same state on an unhandled action', () => {
  expect(data(initialState, { type: '_NULL' })).toMatchSnapshot()
})

it('handles FETCH_GAME_DATA_REQUEST action', () => {
  expect(data(initialState, fetchGameDataRequest())).toMatchSnapshot()
})

it('handles FETCH_GAME_DATA_SUCCESS action', () => {
  const genericGameData = [{ a: 1, b: 2 }, { c: 3, d: 4 }]
  expect(data(initialState, fetchGameDataSuccess(genericGameData))).toMatchSnapshot()
})

it('handles FETCH_GAME_DATA_FAILURE action', () => {
  expect(data(initialState, fetchGameDataFailure())).toMatchSnapshot()
})

// test actions

it('creates a FETCH_GAME_DATA_REQUEST action', () => {
  expect(fetchGameDataRequest()).toMatchSnapshot()
})

it('creates a FETCH_GAME_DATA_SUCCESS action', () => {
  const genericGameData = [{ a: 1, b: 2 }, { c: 3, d: 4 }]
  expect(fetchGameDataSuccess(genericGameData)).toMatchSnapshot()
})

it('creates a FETCH_GAME_DATA_FAILURE action', () => {
  expect(fetchGameDataFailure()).toMatchSnapshot()
})

// test thunks

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

// Figure out how to handle random entries

it('handles fetchMultipleChoiceData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchMultipleChoiceData('set:letters', 4, 2))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchMatchData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchMatchData('set:letters', 4, 2))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchSerialData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchSerialData('set:letters', 4))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchConsecutiveData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchConsecutiveData('set:letters', 5, 4, 3))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchWordData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchWordData('set:letters', 5, 4, 3))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchEquationData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchEquationData('set:sum', 3, 2))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchCrosswordData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchCrosswordData('set:letters', 5, 2))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchTrueOrFalseData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchTrueOrFalseData('set:letters', 3))
  expect(store.getActions()).toMatchSnapshot()
})

it('handles fetchRollingData action', () => {
  const store = mockStore(initialState)
  store.dispatch(fetchRollingData('set:letters', 5, 7, 2))
  expect(store.getActions()).toMatchSnapshot()
})

