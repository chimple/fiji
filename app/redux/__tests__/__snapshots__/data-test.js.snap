// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a FETCH_GAME_DATA_FAILURE action 1`] = `
Object {
  "type": "Fiji/game/FETCH_GAME_DATA_FAILURE",
}
`;

exports[`creates a FETCH_GAME_DATA_REQUEST action 1`] = `
Object {
  "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
}
`;

exports[`creates a FETCH_GAME_DATA_SUCCESS action 1`] = `
Object {
  "gameData": Array [
    Object {
      "a": 1,
      "b": 2,
    },
    Object {
      "c": 3,
      "d": 4,
    },
  ],
  "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
}
`;

exports[`handles FETCH_GAME_DATA_FAILURE action 1`] = `
Object {
  "gameData": Array [],
  "isFetching": false,
}
`;

exports[`handles FETCH_GAME_DATA_REQUEST action 1`] = `
Object {
  "gameData": Array [],
  "isFetching": true,
}
`;

exports[`handles FETCH_GAME_DATA_SUCCESS action 1`] = `
Object {
  "gameData": Array [
    Object {
      "a": 1,
      "b": 2,
    },
    Object {
      "c": 3,
      "d": 4,
    },
  ],
  "isFetching": false,
}
`;

exports[`handles fetchConsecutiveData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "others": Array [
          "160",
          "550",
          "827",
          "918",
        ],
        "serial": Array [
          "321",
          "322",
          "323",
          "324",
          "325",
        ],
      },
      Object {
        "others": Array [
          "619",
          "750",
          "594",
          "103",
        ],
        "serial": Array [
          "237",
          "238",
          "239",
          "240",
          "241",
        ],
      },
      Object {
        "others": Array [
          "136",
          "660",
          "812",
          "842",
        ],
        "serial": Array [
          "494",
          "495",
          "496",
          "497",
          "498",
        ],
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchCrosswordData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "crossword": Array [
          Array [
            "E",
            null,
            null,
            null,
            null,
          ],
          Array [
            "A",
            null,
            null,
            null,
            null,
          ],
          Array [
            "T",
            "I",
            "G",
            "E",
            "R",
          ],
          Array [
            null,
            null,
            null,
            null,
            "A",
          ],
          Array [
            null,
            null,
            null,
            null,
            "T",
          ],
        ],
        "start": Array [
          Array [
            "image",
            Array [
              0,
              0,
            ],
            "down",
          ],
          Array [
            "image",
            Array [
              2,
              0,
            ],
            "lToR",
          ],
          Array [
            "image",
            Array [
              2,
              4,
            ],
            "down",
          ],
        ],
      },
      Object {
        "crossword": Array [
          Array [
            "E",
            null,
            null,
            null,
            null,
          ],
          Array [
            "A",
            null,
            null,
            null,
            null,
          ],
          Array [
            "T",
            "I",
            "G",
            "E",
            "R",
          ],
          Array [
            null,
            null,
            null,
            null,
            "A",
          ],
          Array [
            null,
            null,
            null,
            null,
            "T",
          ],
        ],
        "start": Array [
          Array [
            "image",
            Array [
              0,
              0,
            ],
            "down",
          ],
          Array [
            "image",
            Array [
              2,
              0,
            ],
            "lToR",
          ],
          Array [
            "image",
            Array [
              2,
              4,
            ],
            "down",
          ],
        ],
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchEquationData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "answer": 10,
        "firstNum": 8,
        "op": "+",
        "secondNum": 2,
      },
      Object {
        "answer": 9,
        "firstNum": 5,
        "op": "+",
        "secondNum": 4,
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchMatchData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Array [
        Array [
          "y",
          "y",
        ],
        Array [
          "T",
          "T",
        ],
        Array [
          "c",
          "c",
        ],
        Array [
          "v",
          "v",
        ],
      ],
      Array [
        Array [
          "Q",
          "Q",
        ],
        Array [
          "G",
          "G",
        ],
        Array [
          "T",
          "T",
        ],
        Array [
          "X",
          "X",
        ],
      ],
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchMultipleChoiceData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "answerIndex": 2,
        "choices": Array [
          "C",
          "T",
          "G",
          "X",
        ],
        "question": "G",
      },
      Object {
        "answerIndex": 0,
        "choices": Array [
          "W",
          "Q",
          "Y",
          "J",
        ],
        "question": "W",
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchRollingData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "answer": Array [
          "A",
          "P",
          "P",
          "L",
          "E",
        ],
        "choices": Array [
          Array [
            "X",
            "Y",
            "Z",
            "A",
            "B",
            "C",
            "D",
          ],
          Array [
            "M",
            "N",
            "O",
            "P",
            "Q",
            "R",
            "S",
          ],
          Array [
            "M",
            "N",
            "O",
            "P",
            "Q",
            "R",
            "S",
          ],
          Array [
            "I",
            "J",
            "K",
            "L",
            "M",
            "N",
            "O",
          ],
          Array [
            "B",
            "C",
            "D",
            "E",
            "F",
            "G",
            "H",
          ],
        ],
        "question": "Fruit",
      },
      Object {
        "answer": Array [
          "A",
          "P",
          "P",
          "L",
          "E",
        ],
        "choices": Array [
          Array [
            "X",
            "Y",
            "Z",
            "A",
            "B",
            "C",
            "D",
          ],
          Array [
            "M",
            "N",
            "O",
            "P",
            "Q",
            "R",
            "S",
          ],
          Array [
            "M",
            "N",
            "O",
            "P",
            "Q",
            "R",
            "S",
          ],
          Array [
            "I",
            "J",
            "K",
            "L",
            "M",
            "N",
            "O",
          ],
          Array [
            "B",
            "C",
            "D",
            "E",
            "F",
            "G",
            "H",
          ],
        ],
        "question": "Fruit",
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchSerialData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "answer": 152,
        "serial": Array [
          "147",
          "148",
          "149",
          "150",
          "151",
          "152",
          "153",
          "154",
        ],
      },
      Object {
        "answer": 591,
        "serial": Array [
          "585",
          "586",
          "587",
          "588",
          "589",
          "590",
          "591",
          "592",
        ],
      },
      Object {
        "answer": 778,
        "serial": Array [
          "770",
          "771",
          "772",
          "773",
          "774",
          "775",
          "776",
          "777",
          "778",
          "779",
          "780",
          "781",
          "782",
          "783",
          "784",
          "785",
        ],
      },
      Object {
        "answer": 355,
        "serial": Array [
          "352",
          "353",
          "354",
          "355",
          "356",
          "357",
          "358",
          "359",
          "360",
        ],
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchTrueOrFalseData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "answer": "Fruit",
        "question": "Apple is a:",
        "wrongChoice": "Animal",
      },
      Object {
        "answer": "Fruit",
        "question": "Apple is a:",
        "wrongChoice": "Animal",
      },
      Object {
        "answer": "Fruit",
        "question": "Apple is a:",
        "wrongChoice": "Animal",
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`handles fetchWordData action 1`] = `
Array [
  Object {
    "type": "Fiji/game/FETCH_GAME_DATA_REQUEST",
  },
  Object {
    "gameData": Array [
      Object {
        "others": Array [
          "M",
          "H",
          "T",
          "S",
        ],
        "word": Array [
          "A",
          "P",
          "P",
          "L",
          "E",
        ],
      },
      Object {
        "others": Array [
          "S",
          "J",
          "W",
          "K",
        ],
        "word": Array [
          "A",
          "P",
          "P",
          "L",
          "E",
        ],
      },
      Object {
        "others": Array [
          "T",
          "V",
          "T",
          "I",
        ],
        "word": Array [
          "A",
          "P",
          "P",
          "L",
          "E",
        ],
      },
    ],
    "type": "Fiji/game/FETCH_GAME_DATA_SUCCESS",
  },
]
`;

exports[`returns the same state on an unhandled action 1`] = `
Object {
  "gameData": Array [],
  "isFetching": false,
}
`;
